### ISSUE TITLE
Batchloading and Aliases with parameters

### ISSUE THREAD

**Issue by mheidt** [graphql-java-generator/graphql-maven-plugin-project#73]
Hello Etienne,

I have a nice usecase here:
`a: foo(bar:2)
 b: foo(bar:3)`

I called the dataLoader.load with a batchContext to save bar.
This function is called twice as expected. Once with bar=2 and another time with bar=3.
I create a new batchContext each time.

Unfortunately in the batchloaderEnvironment.keyContexts only bar:2 is seen in alle contexts.
It seems that the second is not taken, as I put a new entry into the loaderQueue with the same key.

What am I doing wrong?

Thanks,
Markus

**Comment by mheidt** [comment]
To be more specific:
In DataLoaderHelper.load there is this part
```

if (cachingEnabled && this.futureCache.containsKey(cacheKey)) {
                this.stats.incrementCacheHitCount();
                return (CompletableFuture)this.futureCache.get(cacheKey);
            } else {
                CompletableFuture<V> future = new CompletableFuture();
                if (batchingEnabled) {
                    this.loaderQueue.add(new DataLoaderHelper.LoaderQueueEntry(key, future, loadContext));
```

I need the cache disabled for batchloader, that have parameters, so that the second call is added to the loaderQueue as well.

I thought, I could be so clever and use my own onDemandDataLoaderRegistry and use something like

```
if (batchLoaderDelegate instanceof BatchLoaderDelegateLocationCustomerCustomersResponseMMImpl){
                        DataLoaderOptions dataLoaderOptions = new DataLoaderOptions();
                        dataLoaderOptions.setCachingEnabled(false);
                        registry.register(batchLoaderDelegate.getName(), DataLoader.newDataLoader(batchLoaderDelegate, dataLoaderOptions));
                    }
```

But that is not enough.
The DataLoaderHelper.dispatch doesn't allow duplicate keys.
It could be, that this a question for graphql-java.

**Comment by mheidt** [comment]
No, I think they will answer, that I should use loadMany, which you would need to provide.
What do you think, @etienne-sf ?

**Comment by etienne-sf** [comment]
Hello,

I have not really understood your use case. 

What means :
´´´
a: foo(bar:2) b: foo(bar:3)
´´´
?

When you say batchContext, one should read keyContext ?

>  I put a new entry into the loaderQueue with the same key

The interest of a dataLoader is to read each keys only once, even if you ask them several times. 
So it's seems natural that you see only once this key. 

Étienne

**Comment by mheidt** [comment]
Maybe this explains it better:

```
{
   product {
     priceOf100: price(amount:100)
     priceOf1000: price(amount:1000)
  }
}
```

in my case, it looks like

```
@Override
    public CompletableFuture<LocationCustomerBusinessPartnerResponseMM> locationCustomerBusinessPartner(DataFetchingEnvironment dataFetchingEnvironment, DataLoader<Long, LocationCustomerBusinessPartnerResponseMM> dataLoader, LocationCustomerMM origin, Long customerType) {
        log.debug("locationCustomerBusinessPartner(dataloader) called");
        Map<String, Object> batchContext = createBatchloaderContext(dataFetchingEnvironment, origin);
        batchContext.put("customerType", customerType);
        return dataLoader.load(origin.getId(), batchContext);
    }
```
Hence the function is called twice for each origin. Twice per key. That can't work, if you don't use loadMany from graphql-java.
 
and debugging confirmed, that
```
 @Override
    public List<LocationCustomerBusinessPartnerResponseMM> batchLoader(List<Long> keys, BatchLoaderEnvironment environment) {
        log.debug("LocationCustomerDF.batchLoader (locationCustomerBusinessPartner) called");
        Map firstMap = (Map)environment.getKeyContexts().values().iterator().next();
        LocationCustomerRequestDTO locationCustomerRequestDTO = new LocationCustomerRequestDTO(
            (CommonFilterDTO)(firstMap.get("filter")));
        locationCustomerRequestDTO.setCustomerType((Long)firstMap.get("customerType"));
        locationCustomerRequestDTO.setLocationCustomerIds(keys);
```
environment.getKeyContexts() size contains only one set of keys...which is correct.
But I don't have any chance to accumulate all "customerTypes"..and rewire it back to the intial calls.
That's what loadMany is for....

**Comment by etienne-sf** [comment]
I don't think so.

loadMany allows you to provide several IDs in one call.

But the objective of the DataLoader is to retrieve only once each item, even if it has been asked several times. So, in your case, there is an issue.

For instance, on this page (https://github.com/graphql/dataloader), it's written:
> After .load() is called once with a given key, the resulting value is cached to eliminate redundant loads.
> ...
> Subsequent calls to .load() with the same key will result in that key not appearing in the keys provided to your batch function


In your case, I would recommand to "change" the key with your own algorithm.

For instance, in this case:
```
{
   product {
     priceOf100: price(amount:100)
     priceOf1000: price(amount:1000)
  }
}
```

Having these keys may do the trick: 
```
dataLoader.load("key1_amount_100");
dataLoader.load("key1_amount_1000"); 
```
Of course, you'll then have to decode your keys afterward

**Comment by mheidt** [comment]
yes, that is my last resort.
But wouldn't it be better if the plugin would generate the boilerplate code?
I would use key + aliasname as dataloader-key.
But I have to take care of the key (ID vs String).
When I am going to try this approach I am curious, if your plugin is generating the batchloader-code if I change the id from ID to String.

**Comment by etienne-sf** [comment]
> But wouldn't it be better if the plugin would generate the boilerplate code?

It is the objective of the plugin, of course
:)
But this use case seems very specific to your use case.


> But I have to take care of the key (ID vs String).

You're right, there is an issue if your ids are ID, then you can not do the trick I propose.

There is a second issue: the plugin currently can't manage alias. 
AFAIU, `priceOf100` and `priceOf1000` are aliases in your request, for the price field, respectively with the amount parameter being 100 then 1000. This is a limit of the POJO generation: once they are generated, the server response must fit with these objects, including field names.
So this request can't be deserialized on client side by the plugin. But perhaps you use it only on server side?

I'm thinking around this, but have no good idea on this subject yet.

A workaround could be to have several prices fields in your product object:
```
type Product {
   name: String
   ...
   # price is the Price per unit
   price: Int
   # price1, price2, price3 (...) allows to have different prices for different amounts
   price1(amount: Int): Int
   price2(amount: Int): Int
   price3(amount: Int): Int
   price4(amount: Int): Int
}
```

Of course, it's far from perfect. But it could do the trick.

**Comment by mheidt** [comment]
Your plugin won't generate Batchloading boilerplate code, if the result type is an Int.
And the purpose was to have one service call in the end.
When using different price fields, I will end up with separate calls.

I disagree, that this is a special use case. Once you begin to think in "graphql" and batchloading, this is the main feature why batchloading was invented, imho.

**Comment by etienne-sf** [comment]
Ok, let me rephrase my understanding:

DataLoader is made to load an object only once for each id, to minimize the number of access to the underlying storage.

My understanding is that you want to load twice that object, to use different parameters when retrieving it. IMHO, this is a special use case, quite different from the DataLoader use.

And the most important is that I don't find any standard way to use a DataLoader, and makes it retrieve to time the same object, based on a secondary attribute.

AFAIU, loadMany won't help.

So the limit is not in the plugin, but in the DataLoader package. 

But if you find a way to make the DataLoader retrieve the Price object two times, based on the request _amount_ parameter, I'll be happy to look if it can be easily embedded in the plugin.

**Comment by mheidt** [comment]
The limitation of the plugin is, that you only allow ID or UUID keys.
That's why I am not able to realize spec-compliant calls.
Now that you took the creation of the onDemandDataLoaderRegistry and the request based creation of the dataloaders, I am able write dataloaders that are using the ExecutionId as keys or something like that.

Don't you consider it a bug, that if you have a batchloaded function with parameters and if you call it with aliases more than once in an object (the spec allows that), you get the same result (the one of the first)?

Aren't you at least be able to throw an exeception, if it is called with an alias?

**Comment by mheidt** [comment]
Etienne, could you provide a mechanism, so that I can add my own wirings and datafetchers?

**Comment by etienne-sf** [comment]
> Don't you consider it a bug, that if you have a batchloaded function with parameters and if you call it with aliases more than once in an object (the spec allows that), you get the same result (the one of the first)?

As stated above, my understanding is that this is the way DataLoader works. So it's not an issue of the plugin, but of DataLoader itself.
From what you've written around, I understand that this can be wrong. But I can't figure exactly how you do it otherwise.

> could you provide a mechanism, so that I can add my own wirings and datafetchers?

I'm thinking around that, to have a fine grained control on what's generated. 

There is already a way, currently, with the [current templates](https://graphql-maven-plugin-project.graphql-java-generator.com/customtemplates.html).
In your case, what would be generated, and what would not be generated ?

From your question, I understand that you would like to override the GraphQLProvider.buildWiring(), to define your own wiring.

Is that correct ?

**Comment by mheidt** [comment]
If you would use a Class, that contains the ID and a payload object as key for the Dataloading, everybody would be able to be far more flexible.
Maybe @Mighty683 would consider using more server code as well if you had more flexibility.

And yes, being able to provide own wiring for certain parts (annoted?) would be swell.

**Comment by etienne-sf** [comment]
It's always the issue with code generator, that it's not possible to manage all possible cases. 

So a good way is to make the generated code open, so that one can override parts of it. As a first step, I'll make the GraphQLProvider method non private, so that they can be overriden.  

Can you precise to me what you'd like to override?


Then I don't see how you can chose that through annotations. Can you provide a sample?

**Comment by mheidt** [comment]
The problem is, that your generated code builds the wiring for everything. I need to opt-out parts of it.
That's what I meant with annotation. If I annotate a field, you won't generate the wiring for it.
And then we have several options to hook into it. For example:
I would give GraphQLProvider an interface, so that the "customer" is able to @primary his or her own GraphQLProvider.
I would inherit from your GraphQLProvider and override it.
You shouldn't make a hook in buildWiring() so that we are able to add our own wirings by overriding that function.

`RuntimeWiring.builder createCustomWiring(RuntimeWiring.builder builder){ return builder; } `

**Comment by etienne-sf** [comment]
Ok, that seems nice. 

If you're ok, we'll try that: 
I'll create the annotation and the hook point (with an interface like your proposed). You'll test it on the snapshot version, before I release it.

**Comment by mheidt** [comment]
Sure, gladly do. You tell me when?

**Comment by etienne-sf** [comment]
I'm starting working on this one. And I think that putting an annotation on a field is complex and error-prone.
How to be sure that the wiring is set for all the fields that have this annotation ? How to be sure that custom wiring are defined only for fields that have this annotation ? ...

I think there is a better way to do this :

Let the `GraphQLProvider.buildWiring()` define every wiring. 
But for each wiring, it would cal a local protected method. Then, to change the default wiring, you just have to define a Primary GraphQLProvider bean of your own, an override one or more of these methods.

**Comment by etienne-sf** [comment]
Hello,

I updated the `GraphQLProvider` methods, so that they are protected. You can now:
* Create a `CustomGraphQLProvider` Spring bean of your own, 
* Mark it as _@Primary_ so that it replaces the default one
* Inherit from `GraphQLProvider`
* Override the method you would like to change, to change its behavior.

In your case, you want to override the DataFetcher definition for one field of one type.

To do this, the idea is the same, but with the `GraphQLDataFetchers` class:
* Create a `CustomGraphQLDataFetchers` Spring bean of your own, 
* Mark it as _@Primary_ so that it replaces the default one
* Inherit from `GraphQLDataFetchers`
* Override the method you would like to change, to change the relevant data fetcher. 

I add a JUnit test to check that, in the forum sample. 

You can see an example of such a `CustomGraphQLDataFetchers` Spring bean here :
https://github.com/graphql-java-generator/graphql-maven-plugin-project/blob/master/graphql-maven-plugin-samples/graphql-maven-plugin-samples-Forum-server/src/main/java/org/forum/server/specific_code/CustomGraphQLDataFetchers.java

Please note that overriding `GraphQLDataFetchers` with a `CustomGraphQLDataFetchers` Spring bean is already possible with the 1.12.3 version.

Please let me know if it works Ok for you.

Etienne

**Comment by mheidt** [comment]
When I tried to build the complete project, the samples-Forum-Server fails.
CustomGraphQLDataFetchers. graphUtils is not public in  ..GraphQLDataFetchers..cannot be accessed from outside package

**Comment by etienne-sf** [comment]
Are building the whole projet, from the last snapshot ?
If Yes, did you do a proper `mvn clean install` ?
(If not, the last correction may not be regenerated)

Or are you trying to use the `CustomGraphQLDataFetchers`, on the 1.12.3 release ?

In the later case, yes, I just commited a correction for this one.
Sorry for that, I forgot this point.

If this is the ase, you have two options there:
1) Copy/paste theses lines into the `CustomGraphQLDataFetchers` class:
```
	/** The logger for this instance */
	protected Logger logger = LoggerFactory.getLogger(GraphQLDataFetchers.class);

	@Resource
	protected DataFetchersDelegateQueryType dataFetchersDelegateQueryType;

	@Resource
	protected DataFetchersDelegateMutationType dataFetchersDelegateMutationType;

	@Resource
	protected DataFetchersDelegateSubscriptionType dataFetchersDelegateSubscriptionType;

	@Resource
	protected DataFetchersDelegateBoard dataFetchersDelegateBoard;

	@Resource
	protected DataFetchersDelegateTopic dataFetchersDelegateTopic;

	@Resource
	protected DataFetchersDelegatePost dataFetchersDelegatePost;

	@Resource
	protected DataFetchersDelegateMember dataFetchersDelegateMember;

	@Resource
	protected GraphqlUtils graphqlUtils;
```
2) Use the Snapshot version, in order to use the corrected version of GraphQLDataFetchers

**Comment by mheidt** [comment]
I tried to get local-snapshot running. But now the OAuth-auth-server process times out after 180sec.
mvn clean install -DskipTests

Now I give 1.12.3 a try

**Comment by mheidt** [comment]
Did you change anything, so that I can hook into the buildwiring? I don't see any changes in that area.
Actually, I don't want to copy/paste the 200 lines of generated wirings only because I have to modify 10 wirings in my CustomGraphQLProvider. (and compare those each time I regenerated the schema)

**Comment by etienne-sf** [comment]
> I tried to get local-snapshot running. But now the OAuth-auth-server process times out after 180sec.
> mvn clean install -DskipTests

I could not reproduce that one (despite my slow PC)

Should I understand that you're now able to override the wiring, as you expected ?

**Comment by mheidt** [comment]
YES. Thanks...Actually for this Alias issue, I didn't need to override the wiring.

A CustomGraphQLDataFetchers and two new DataFetchers interfaces, plus a new BatchLoaderDelegateImpl and overriding the function, that is calling getDataLoader("myNewBatchLoader"), was enough.

But I will try to get rid of those intermittent classes I created to convince the plugin to generate the batchloaders. 
That's where I need the overridden wirings.

But first I  would be glad, if you could build 1.12.4.
We can close this task...A BIG STEP FORWARD. WELL DONE!

**Comment by etienne-sf** [comment]
Thanks Markus

Now I need to properly document all this, to make it clear to new comers.

**Comment by mheidt** [comment]
I have some issues, using it. Maybe this helps to get a proper documentation.
I get the Alias thing working, when I use your generator like follows

```
locationCustomers(customerType: Long): LocationCustomersResponseMM
type LocationCustomersResponseMM {
  id: ID
  values: [LocationCustomerMM!]
}
```

But the ultimate goal would be this
` locationCustomers(customerType: Long): [LocationCustomerMM!]`

Hence I stole the dataFetchersDelegateLocationMMLocationCustomersWithDataLoader from GraphQLDataFetchers from the first build and copied it to CustomGraphQLDataFetchers. 

Then I had to override 

```
@Component
@Primary
public class CustomGraphQLProvider extends GraphQLProvider {

    @Autowired
    CustomGraphQLDataFetchers graphQLDataFetchers;

        @Override
        protected TypeRuntimeWiring.Builder addWiringForLocationMM(TypeRuntimeWiring.Builder typeWiring) {
```
For now, I copy the original but replace
`typeWiring.dataFetcher("locationCustomers", graphQLDataFetchers.dataFetchersDelegateLocationMMLocationCustomersWithDataLoaderEx());`

So the issue is, that unfortunatly the original
`protected TypeRuntimeWiring.Builder addWiringForLocationMM(TypeRuntimeWiring.Builder typeWiring) {`
in GraphQLProvider is called first. And my overriding later.

That's why dataFetchersDelegateLocationMMLocationCustomers (without DataLoader) is called.

So maybe, you should provide us a two step mechanism...first collect and then write into typeWiring. Unfortunatley graphql -java doesn't allow to remove datafetcher.

**Comment by etienne-sf** [comment]
> So the issue is, that unfortunatly the original
> `protected TypeRuntimeWiring.Builder addWiringForLocationMM(TypeRuntimeWiring.Builder typeWiring) {`
> in GraphQLProvider is called first. And my overriding later.

I don't understand this one : this should not happen. If you override a method, the original one should not be called. This is a core java behavior.

Or perhaps the two GraphQLProvider are executed ?

**Comment by mheidt** [comment]
Yes, I was setting breakpoints and both provider functions were called.

**Comment by mheidt** [comment]
It's because, it's called from a @PostConstruct

**Comment by etienne-sf** [comment]
Yes, I've seen an issue there. 
When overriding the GraphQLProvider, I get an error with the GraphQL bean being defined twice.

I'll work on this one. 

Perhaps by extracting all the wiring stuff into the GraphQLDataFetchers class. I'll think around that.

**Comment by etienne-sf** [comment]
Hello,

I've split the content of the GraphQLProvider into two parts:
* The bean definitions have been moved into the GraphQLMain class
* The type wiring have been moved into the new GraphQLWiring class

I added a CustomGraphQLWiring class in the Forum server sample, to be sure that overriding of the GraphQLWiring class works Ok.

This should correct the issue you had.

**Comment by mheidt** [comment]
Yes! Perfect.
My intermediate class is gone!

I would be ready for 1.14, if you are :)
